### PR TITLE
Feature/while op sentiment analysis

### LINKED ISSUE #5861: Dynamic RNN API
Currently, we only have `while_op`, `lod_tensor_to_array_op` and `array_to_lod_tensor_op`. It's too hard for users to build a complete dynamic RNN with these basic components. So we need to warp them together in Python as a new API.

### PR DESCRIPTION
**Pull request by reyoung**
fixes #5861

### CODE REVIEW

File: python/paddle/v2/fluid/layers.py [hunk 1]
```diff
@@ -1838,3 +1843,209 @@ def __call__(self):
                     main_program=self.helper.main_program,
                     startup_program=self.helper.startup_program))
         return rlist
+
+
+class DynamicRNN(object):
+    BEFORE_RNN = 0
+    IN_RNN = 1
+    AFTER_RNN = 2
```
Comment (QiJune) [inline python/paddle/v2/fluid/layers.py]: So, can we use nested dynamic RNN? Should the status be IN_RNN and OUT_RNN?

File: python/paddle/v2/fluid/layers.py [hunk 2]
```diff
@@ -1838,3 +1843,209 @@ def __call__(self):
...
+                inputs={"X": x},
+                outputs={"Out": self.lod_rank_table})
+            self.max_seq_len = parent_block.create_var(
+                name=unique_name('dynamic_rnn_max_seq_len'), dtype='int64')
+            self.max_seq_len.stop_gradient = False
+            parent_block.append_op(
+                type='max_sequence_len',
+                inputs={'RankTable': self.lod_rank_table},
+                outputs={"Out": self.max_seq_len})
+            self.cond.stop_gradient = True
```
Comment (QiJune) [inline python/paddle/v2/fluid/layers.py]: Why stop_gradient here is True

File: python/paddle/v2/fluid/layers.py [hunk 3]
```diff
@@ -1838,3 +1843,209 @@ def __call__(self):
...
+        self.status = DynamicRNN.BEFORE_RNN
+        self.lod_rank_table = None
+        self.max_seq_len = None
+        self.step_idx = None
+        self.zero_idx = fill_constant(shape=[1], value=0, dtype='int64')
+        self.mem_dict = dict()
+        self.output_array = []
+        self.outputs = []
+        self.cond = self.helper.create_tmp_variable(dtype='bool')
+        self.cond.stop_gradient = False
```
Comment (QiJune) [inline python/paddle/v2/fluid/layers.py]: I am little confused about these two concepts, stop_gradient and no_gradient. Are these two the same?

File: python/paddle/v2/fluid/layers.py (hunk 2 above)
Comment (reyoung) [reply python/paddle/v2/fluid/layers.py]: It is no need to calculate gradient of less than

File: python/paddle/v2/fluid/layers.py (hunk 3 above)
Comment (reyoung) [reply python/paddle/v2/fluid/layers.py]: `no_grad_set` is a parameter of `backward(...)` method. stop_gradient is an attribute of Python Variable.

The variables whose stop_gradient=True will be passed to `backward(no_grad_set=...)`

File: python/paddle/v2/fluid/layers.py [hunk 4]
```diff
@@ -1838,3 +1843,209 @@ def __call__(self):
...
+
+    @contextlib.contextmanager
+    def block(self):
+        if self.status != DynamicRNN.BEFORE_RNN:
+            raise ValueError("rnn.block() can only be invoke once")
+        self.step_idx = fill_constant(shape=[1], dtype='int64', value=0)
+        self.step_idx.stop_gradient = False
+        self.status = DynamicRNN.IN_RNN
+        with self.while_op.block():
+            yield
```
Comment (QiJune) [inline python/paddle/v2/fluid/layers.py]: Why yield here? Not before line 1922

File: python/paddle/v2/fluid/tests/test_dyn_rnn.py [hunk 5]
```diff
@@ -82,6 +83,48 @@ def test_plain_while_op(self):
...
+            rnn = fluid.layers.DynamicRNN()
+
+            with rnn.block():
+                in_ = rnn.step_input(sent_emb)
+                mem = rnn.memory(shape=[100], dtype='float32')
+                out_ = fluid.layers.fc(input=[in_, mem], size=100, act='tanh')
+                rnn.update_memory(mem, out_)
+                rnn.output(out_)
+
+            last = fluid.layers.sequence_pool(input=rnn(), pool_type='last')
```
Comment (jacquesqiao) [inline python/paddle/v2/fluid/tests/test_dyn_rnn.py]: rnn() 改成  rnn.out  是不是更容易理解点

Comment (reyoung) [reply python/paddle/v2/fluid/tests/test_dyn_rnn.py]: Just follow the API design on https://github.com/PaddlePaddle/talks/blob/develop/paddle-gtc-china.pdf

**Pull request merged**